Add a progress indicator for `cargo clean`

theo-lw opened the pull request:
Closes https://github.com/rust-lang/cargo/issues/9981.

Cleaning the entire `target` directory:

[![Cleaning `target`](https://i.gyazo.com/e32536fe5538cf5a394d777a9d838890.gif)](https://gyazo.com/e32536fe5538cf5a394d777a9d838890)

Cleaning some packages:

[![Cleaning a few packages](https://i.gyazo.com/860f1833523afc0629d5547d97c146c2.gif)](https://gyazo.com/860f1833523afc0629d5547d97c146c2)

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @ehuss (or someone else) soon.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

weihanglo reviewed (commented):
The progress bar seems disappear when `--verbose` is specified with `-p`. I guess it is replaced by then output of `Removing ...`. We can tick the progress bar after each `rm_rf` but if the removing is stuck the progress bar would still disappear for a while.

theo-lw commented on src/cargo/ops/cargo_clean.rs:
```diff
@@ -231,6 +235,25 @@ fn rm_rf_glob(pattern: &Path, config: &Config) -> CargoResult<()> {
+                paths::remove_dir(entry.path())?;
+            } else {
+                paths::remove_file(entry.path())?;
+            }
```
Oops, updated so the verbose `Removing ...` is kept.

theo-lw commented on src/cargo/ops/cargo_clean.rs:
```diff
@@ -133,8 +135,10 @@ pub fn clean(ws: &Workspace<'_>, opts: &CleanOptions<'_>) -> CargoResult<()> {
+    let mut progress = Progress::with_style("Cleaning", ProgressStyle::Ratio, config);
+    for (pkg_idx, pkg) in packages.iter().enumerate() {
         let pkg_dir = format!("{}-*", pkg.name());
+        progress.tick_now(pkg_idx + 1, packages.len(), &format!(": {}", pkg.name()))?;
```
It would. I added a message listing of the # of files/folders cleaned so users can see that it isn't stuck.

theo-lw commented:
Fixed the issue of `Removing...` not being shown in verbose mode:

[![Image from Gyazo](https://i.gyazo.com/f80a5e159e7bda7331731589d24bc896.gif)](https://gyazo.com/f80a5e159e7bda7331731589d24bc896)

Fixed the issue of the progress bar disappearing in verbose mode:

[![Image from Gyazo](https://i.gyazo.com/425a99ef575c3bca729f86b42a7b4ec4.gif)](https://gyazo.com/425a99ef575c3bca729f86b42a7b4ec4)

weihanglo commented on src/cargo/ops/cargo_clean.rs:
```diff
@@ -34,7 +34,7 @@ pub fn clean(ws: &Workspace<'_>, opts: &CleanOptions<'_>) -> CargoResult<()> {
     if opts.doc {
         target_dir = target_dir.join("doc");
-        return rm_rf(&target_dir.into_path_unlocked(), config);
+        return clean_entire_folder(&target_dir.into_path_unlocked(), &config);
```
NIT: need not to borrow
```suggestion
        return clean_entire_folder(&target_dir.into_path_unlocked(), config);
```

weihanglo commented on src/cargo/ops/cargo_clean.rs:
```diff
@@ -53,7 +53,7 @@ pub fn clean(ws: &Workspace<'_>, opts: &CleanOptions<'_>) -> CargoResult<()> {
     // blow it all away anyway.
     if opts.spec.is_empty() {
-        return rm_rf(&target_dir.into_path_unlocked(), config);
+        return clean_entire_folder(&target_dir.into_path_unlocked(), &config);
```
ditto
```suggestion
        return clean_entire_folder(&target_dir.into_path_unlocked(), &config);
```

weihanglo reviewed (approved):
Awesome. I think it's good to move forward and wait the review from maintainers. 🎉

alexcrichton commented on src/cargo/ops/cargo_clean.rs:
```diff
@@ -220,29 +222,134 @@ fn escape_glob_path(pattern: &Path) -> CargoResult<String> {
 }
+
+fn clean_entire_folder(path: &Path, config: &Config) -> CargoResult<()> {
+    let num_paths = walkdir::WalkDir::new(path).into_iter().count();
```
This naively seems like it would have a large impact and slow things down because we have to walk everything before actually removing everything. I don't have a great sense for the practical impact of this though.

alexcrichton replied:
An alternative to this I think would be to perform the walk once, collect a list of all paths to remove, then remove them all in a loop. That would mean we only have to walk once and should be similar in performance I think.

alexcrichton commented on src/cargo/ops/cargo_clean.rs:
```diff
@@ -220,29 +222,134 @@ fn escape_glob_path(pattern: &Path) -> CargoResult<String> {
+fn rm_rf_glob(
+    pattern: &Path,
+    config: &Config,
+    progress: &mut impl CleaningProgressBar,
```
I think it'd be ok to change this from `impl Cleaning...` to `dyn Cleaning...` since we don't need the monomorphization aspect here per se

alexcrichton reviewed (commented):
Seems reasonable to me, thanks for this!

ehuss commented:
Just to echo some of Alex's comments, can you do some benchmarking and report if there is a performance hit with this? I think it would be good to test with some medium-ish sized projects, and on different systems (windows in particular has some slow fs behaviors).  A small hit is fine, but if this makes `cargo clean` twice as slow, that might be concerning.

theo-lw commented:
I'll do some benchmarking :) didn't realize that Windows had some peculiarities

theo-lw commented:
Running `cargo clean` on rust-analyzer on Windows: 

Before these changes:

![Screen Shot 2022-01-05 at 1 16 10 AM](https://user-images.githubusercontent.com/28904678/148246803-c2ee2077-c4d4-4556-964e-70d0f981c437.png)

After these changes:

![Screen Shot 2022-01-05 at 1 22 50 AM](https://user-images.githubusercontent.com/28904678/148246858-4105a496-20ca-458d-a8ec-a2f4d6be0ad7.png)

So it looks like there is a performance regression of ~12% on Windows. 

Running `cargo clean` on rust-analyzer on my mac:

Before these changes:

![Screen Shot 2022-01-05 at 12 08 46 PM](https://user-images.githubusercontent.com/28904678/148262309-a7d84ec2-0bf3-41ef-8b59-bda7a6f08f1a.png)

After these changes:

![Screen Shot 2022-01-05 at 12 25 35 PM](https://user-images.githubusercontent.com/28904678/148262276-043aea41-10db-422a-a059-37fb4b074b59.png)

Performance doesn't have seemed to change much on my mac.

I'll experiment with collecting the paths into a vector.

ehuss commented:
I did some testing on various systems and different project sizes, and got roughly the same performance hit that you mentioned (ranging from 0 to 10%).  The small performance hit is a bit unfortunate, but overall I don't think will be noticed much.  I'm a little concerned for using cargo over a slow network connection, as I think this could tick quite frequently.  I think there can possibly be followup in the future if that turns into a serious concern.  Another possibility is to explore deleting files in parallel.

@rust-lang/cargo I'm going to just do a quick fcp check since this changes the output of a stable command, and there is a small performance hit.

@rfcbot fcp merge

weihanglo commented:
Just posted the possible extensions as a new issue (https://github.com/rust-lang/cargo/issues/10552). I am going to merge this PR.

@bors r+

bors commented:
:pushpin: Commit 60cfe7ef51a9abea164037b6a0939e493441bcb3 has been approved by `weihanglo`

<!-- @bors r=weihanglo 60cfe7ef51a9abea164037b6a0939e493441bcb3 -->
<!-- homu: {"type":"Approved","sha":"60cfe7ef51a9abea164037b6a0939e493441bcb3","approver":"weihanglo"} -->

bors commented:
:hourglass: Testing commit 60cfe7ef51a9abea164037b6a0939e493441bcb3 with merge f51e799636fcba6aeb98dc2ca7e440ecd9afe909...
<!-- homu: {"type":"BuildStarted","head_sha":"60cfe7ef51a9abea164037b6a0939e493441bcb3","merge_sha":"f51e799636fcba6aeb98dc2ca7e440ecd9afe909"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/5959869789?check_suite_focus=true)
Approved by: weihanglo
Pushing f51e799636fcba6aeb98dc2ca7e440ecd9afe909 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"weihanglo","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/5959869789?check_suite_focus=true"},"merge_sha":"f51e799636fcba6aeb98dc2ca7e440ecd9afe909"} -->
